src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1505,12 +1636,179 @@ namespace Microsoft.FSharp.Core
                     override iec.Equals(x:obj,y:obj) = GenericEqualityObj true iec (x,y)  // ER Semantics
                     override iec.GetHashCode(x:obj) = raise (InvalidOperationException (SR.GetString(SR.notUsedForHashing))) }
 
+            let canUseDefaultEqualityComparer (ty:Type) =
+                let processed = System.Collections.Generic.HashSet ()
+
+                let rec recurse idx (types:array<Type>) =
+                    if idx = types.Length then true
+                    else
+                        let ty = get types idx
+                        if not (processed.Add ty) then
+                            recurse (idx+1) types
+                        else
+                            let isValidGenericType ifNotType fullname =
+                                if not (ty.IsGenericType && ty.GetGenericTypeDefinition().FullName.Equals fullname)
+                                then ifNotType
+                                else recurse 0 (ty.GetGenericArguments ())
+                    
+                            // avoid any types that need special handling in GenericEqualityObj
+                            true 
+                            && ty.IsSealed  // covers enum and value types
+                            && not ty.IsArray
+                            && not (ty.Equals typeof<String>)
+                            && not (ty.Equals typeof<Decimal>)
+                            && not (ty.Equals typeof<float>)
+                            && not (ty.Equals typeof<float32>)
+                            && isValidGenericType true "System.Nullable`1"
+                            && not (typeof<IStructuralEquatable>.IsAssignableFrom ty
+                                    && not (false
```

Comment:
wonky formatting here

Reply:
Using `true && ...` and `false || ...` is a bit odd, just remove

Reply:
The wonky formatting is what is aligned with - i.e what the `not`, `&&` or `||` corresponds to. I find it _much_ easier to read than any other formatting - but happy to change it whatever you want.

The `true` at the start of `&&` and the `false` at the start of `||` allow you to comment out any of the corresponding lines without changing any other lines - the compiler is smart enough to remove them - but if you don't like 'em I'll get rid of them.